File: java/CatalogReal.java
package videoprokat;

import java.text.Collator;
import java.text.Normalizer;
import java.util.ArrayList;
import java.util.Comparator;
import java.util.HashMap;
import java.util.List;
import java.util.Locale;
import java.util.Map;

/**
 * Реализация класса Catalog.
 * Содержит конкретную реализацию всех методов абстрактного класса Catalog.
 */
public class CatalogReal extends Catalog {

    private static final Locale RU = new Locale("ru", "RU");
    private static final Collator COLLATOR = Collator.getInstance(RU);

    static {
        COLLATOR.setStrength(Collator.PRIMARY); // нечувствительно к регистру/акцентам
    }

    /**
     * Конструктор по умолчанию.
     * Инициализирует список носителей.
     */
    public CatalogReal() {
        this.items = new ArrayList<>();
    }

    @Override
    public void addItem(VideoCarrier item) {
        if (item != null) {
            items.add(item);
        }
    }

    @Override
    public VideoCarrier findItemByNumber(int number) {
        for (VideoCarrier item : items) {
            if (item != null && item.getInventoryNumber() == number) {
                return item;
            }
        }
        return null;
    }

    @Override
    public List<VideoCarrier> findItemsByTitle(String title) {
        List<VideoCarrier> result = new ArrayList<>();
        for (VideoCarrier item : items) {
            if (item != null && containsRu(item.getTitle(), title)) {
                result.add(item);
            }
        }
        return result;
    }

    @Override
    public List<VideoCarrier> getAvailableItems() {
        List<VideoCarrier> result = new ArrayList<>();
        for (VideoCarrier item : items) {
            if (item != null && item.isAvailable()) {
                result.add(item);
            }
        }
        return result;
    }

    @Override
    public List<VideoCarrier> findItemsByGenre(String genre) {
        List<VideoCarrier> result = new ArrayList<>();
        for (VideoCarrier item : items) {
            if (item != null && containsRu(item.getGenre(), genre)) {
                result.add(item);
            }
        }
        return result;
    }

    @Override
    public List<VideoCarrier> findItemsByDirector(String director) {
        List<VideoCarrier> result = new ArrayList<>();
        for (VideoCarrier item : items) {
            if (item != null && containsRu(item.getDirector(), director)) {
                result.add(item);
            }
        }
        return result;
    }

    @Override
    public List<VideoCarrier> findItemsByYear(int year) {
        List<VideoCarrier> result = new ArrayList<>();
        for (VideoCarrier item : items) {
            if (item != null && item.getReleaseYear() == year) {
                result.add(item);
            }
        }
        return result;
    }

    @Override
    public List<VideoCarrier> findItemsByCarrierType(String carrierType) {
        List<VideoCarrier> result = new ArrayList<>();
        for (VideoCarrier item : items) {
            if (item != null && equalsRu(item.getCarrierType(), carrierType)) {
                result.add(item);
            }
        }
        return result;
    }

    @Override
    public List<VideoCarrier> findItemsByAgeRating(String ageRating) {
        List<VideoCarrier> result = new ArrayList<>();
        for (VideoCarrier item : items) {
            if (item != null && equalsRu(item.getAgeRating(), ageRating)) {
                result.add(item);
            }
        }
        return result;
    }

    @Override
    public List<VideoCarrier> getAllItems() {
        return new ArrayList<>(items);
    }

    @Override
    public Map<String, Integer> getStatistics() {
        Map<String, Integer> stats = new HashMap<>();
        int available = 0, rented = 0, maintenance = 0, writtenOff = 0;
        
        for (VideoCarrier item : items) {
            if (item != null) {
                String status = item.getStatus();
                if ("available".equals(status)) available++;
                else if ("rented".equals(status)) rented++;
                else if ("maintenance".equals(status)) maintenance++;
                else if ("written_off".equals(status)) writtenOff++;
            }
        }
        
        stats.put("total", items.size());
        stats.put("available", available);
        stats.put("rented", rented);
        stats.put("maintenance", maintenance);
        stats.put("written_off", writtenOff);
        
        return stats;
    }

    @Override
    public List<VideoCarrier> getTopRentedItems(int limit) {
        List<VideoCarrier> sorted = new ArrayList<>(items);
        sorted.sort(new Comparator<VideoCarrier>() {
            @Override
            public int compare(VideoCarrier v1, VideoCarrier v2) {
                return Integer.compare(v2.getTotalRentals(), v1.getTotalRentals());
            }
        });
        
        if (limit > sorted.size()) {
            limit = sorted.size();
        }
        
        return sorted.subList(0, limit);
    }

    private String normalizeRu(String s) {
        if (s == null) return "";
        // Простая нормализация - переводим в нижний регистр
        return s.toLowerCase(RU);
    }

    private boolean containsRu(String haystack, String needle) {
        String n = normalizeRu(needle == null ? "" : needle.trim());
        if (n.isEmpty()) return true; // пустой запрос — считаем совпадением
        if (haystack == null) return false;
        String h = normalizeRu(haystack);
        return h.contains(n); // подстрока, регистронезависимо, работает и на 1 символ
    }

    private boolean equalsRu(String a, String b) {
        return COLLATOR.compare(a == null ? "" : a, b == null ? "" : b) == 0;
    }
}

File: cpp/CatalogReal.cpp
#include "CatalogReal.hpp"
#include "VideoCarrier.hpp"
#include <algorithm>
#include <unordered_map>
#include <cwctype>

#ifdef _WIN32
#include <windows.h>
#else
#include <codecvt>
#include <locale>
#endif

CatalogReal::CatalogReal()
{
    // items инициализируется автоматически
}

CatalogReal::~CatalogReal()
{
    // В C++ не удаляем items, так как они принадлежат другим объектам
}

void CatalogReal::addItem(VideoCarrier* item)
{
    if (item != nullptr) {
        items.push_back(item);
    }
}

VideoCarrier* CatalogReal::findItemByNumber(int number)
{
    for (VideoCarrier* item : items) {
        if (item != nullptr && item->getInventoryNumber() == number) {
            return item;
        }
    }
    return nullptr;
}

std::vector<VideoCarrier*> CatalogReal::findItemsByTitle(const std::string& title)
{
    std::vector<VideoCarrier*> result;
    for (VideoCarrier* item : items) {
        if (item != nullptr && containsIgnoreCaseUtf8(item->getTitle(), title)) {
            result.push_back(item);
        }
    }
    return result;
}

std::vector<VideoCarrier*> CatalogReal::getAvailableItems()
{
    std::vector<VideoCarrier*> result;
    for (VideoCarrier* item : items) {
        if (item != nullptr && item->isAvailable()) {
            result.push_back(item);
        }
    }
    return result;
}

std::vector<VideoCarrier*> CatalogReal::findItemsByGenre(const std::string& genre)
{
    std::vector<VideoCarrier*> result;
    for (VideoCarrier* item : items) {
        if (item != nullptr) {
            if (equalsIgnoreCaseUtf8(item->getGenre(), genre)) {
                result.push_back(item);
            }
        }
    }
    return result;
}

std::vector<VideoCarrier*> CatalogReal::findItemsByDirector(const std::string& director)
{
    std::vector<VideoCarrier*> result;
    for (VideoCarrier* item : items) {
        if (item != nullptr) {
            if (containsIgnoreCaseUtf8(item->getDirector(), director)) {
                result.push_back(item);
            }
        }
    }
    return result;
}

std::vector<VideoCarrier*> CatalogReal::findItemsByYear(int year)
{
    std::vector<VideoCarrier*> result;
    for (VideoCarrier* item : items) {
        if (item != nullptr && item->getReleaseYear() == year) {
            result.push_back(item);
        }
    }
    return result;
}

std::vector<VideoCarrier*> CatalogReal::findItemsByCarrierType(const std::string& carrierType)
{
    std::vector<VideoCarrier*> result;
    for (VideoCarrier* item : items) {
        if (item != nullptr) {
            if (equalsIgnoreCaseUtf8(item->getCarrierType(), carrierType)) {
                result.push_back(item);
            }
        }
    }
    return result;
}

std::vector<VideoCarrier*> CatalogReal::findItemsByAgeRating(const std::string& ageRating)
{
    std::vector<VideoCarrier*> result;
    for (VideoCarrier* item : items) {
        if (item != nullptr && equalsIgnoreCaseUtf8(item->getAgeRating(), ageRating)) {
            result.push_back(item);
        }
    }
    return result;
}

std::vector<VideoCarrier*> CatalogReal::getAllItems()
{
    std::vector<VideoCarrier*> result;
    result.reserve(items.size());
    for (VideoCarrier* item : items) {
        result.push_back(item);
    }
    return result;
}

std::map<std::string, int> CatalogReal::getStatistics()
{
    std::map<std::string, int> stats;
    int available = 0, rented = 0, maintenance = 0, writtenOff = 0;
    
    for (VideoCarrier* item : items) {
        if (item != nullptr) {
            std::string status = item->getStatus();
            if (status == "available") available++;
            else if (status == "rented") rented++;
            else if (status == "maintenance") maintenance++;
            else if (status == "written_off") writtenOff++;
        }
    }
    
    stats["total"] = static_cast<int>(items.size());
    stats["available"] = available;
    stats["rented"] = rented;
    stats["maintenance"] = maintenance;
    stats["written_off"] = writtenOff;
    
    return stats;
}

std::vector<VideoCarrier*> CatalogReal::getTopRentedItems(int limit)
{
    std::vector<VideoCarrier*> sorted = items;
    std::sort(sorted.begin(), sorted.end(), 
        [](VideoCarrier* v1, VideoCarrier* v2) {
            if (v1 && v2) {
                return v2->getTotalRentals() < v1->getTotalRentals();
            }
            return false;
        });
    
    if (limit > static_cast<int>(sorted.size())) {
        limit = static_cast<int>(sorted.size());
    }
    
    std::vector<VideoCarrier*> result;
    result.reserve(limit);
    for (int i = 0; i < limit; ++i) {
        result.push_back(sorted[i]);
    }
    
    return result;
}

// Приведение UTF-8 строки к нижнему регистру через широкие символы (для корректной кириллицы)
static std::wstring toLowerWide(const std::string& s) {
    if (s.empty()) return std::wstring();

#ifdef _WIN32
    int len = MultiByteToWideChar(CP_UTF8, 0, s.c_str(), -1, nullptr, 0);
    if (len <= 1) return std::wstring();
    std::wstring wide(static_cast<size_t>(len - 1), L'\0');
    MultiByteToWideChar(CP_UTF8, 0, s.c_str(), -1, &wide[0], len);
#else
    std::wstring_convert<std::codecvt_utf8_utf16<wchar_t>> conv;
    std::wstring wide = conv.from_bytes(s);
#endif

    for (auto& ch : wide) {
        ch = static_cast<wchar_t>(::towlower(ch));
    }
    return wide;
}

bool CatalogReal::containsIgnoreCaseUtf8(const std::string& str, const std::string& substr)
{
    std::wstring h = toLowerWide(str);
    std::wstring n = toLowerWide(substr);
    if (n.empty()) return true; // пустой запрос — все совпадения
    return h.find(n) != std::wstring::npos;
}

bool CatalogReal::equalsIgnoreCaseUtf8(const std::string& lhs, const std::string& rhs)
{
    return toLowerWide(lhs) == toLowerWide(rhs);
}

// Старый toLowerSimple больше не используется, оставлен для совместимости
std::string CatalogReal::toLowerSimple(const std::string& str)
{
    return str;
}


